Make websocket close include the status code

davidot opened the pull request:
In the websocket protocol the first two bytes of the payload in a close packet indicate the status code.

:warning: This code is not backwards compatible, as I could not find how that is handled in this project. In the most extreme case we can use a std::enable_if on the onclose handler which then puts the status code in front of the message again to ensure any user code behaves the exact same.

Fixes: #545 
And also the third commit fixes #549 (at the very least the simple case)

gittiver commented:
HI davidot, could you update to the newest main branch? It's just for macOS build, which does not work anymore on macos11?

gittiver commented on tests/unittest.cpp:
```diff
@@ -2809,6 +2822,186 @@ TEST_CASE("websocket")
+        // Close message with, len = 2, status code = 1000
+        char close_message[5]("\x88\x02\x03\xE8");
+
+        c.send(asio::buffer(close_message, 10));
```
size of close_message is 5 here?

davidot commented on include/crow/websocket.h:
```diff
@@ -40,7 +40,7 @@ namespace crow // NOTE: Already documented in "crow/app.h"
             virtual void send_ping(std::string msg) = 0;
             virtual void send_pong(std::string msg) = 0;
-            virtual void close(std::string const& msg = "quit") = 0;
+            virtual void close(std::string const& msg = "quit", uint16_t status_code = 1000u) = 0;
```
Maybe but there are not limited to some fixed set as far as I can see.
Would you want to use an (non class) enum just to provide the default "recommended" codes?

gittiver replied:
for the values, yes. I would like to have the default codes as enumeration.

davidot commented on include/crow/websocket.h:
```diff
@@ -683,7 +701,7 @@ namespace crow // NOTE: Already documented in "crow/app.h"
                 if (!is_close_handler_called_)
                     if (close_handler_)
-                        close_handler_(*this, "uncleanly");
+                        close_handler_(*this, "uncleanly", 0);
```
Good point, will use the appropriate status code from the spec

davidot commented:
Rebased on master, added enum (non-class one, can change that) and fixed 0 codes.

CI failure is a already present unit test, although it is not failing on master? It is passing for me locally

gittiver commented:
Restarted the ci jobs, seems GitHub had an hiccup, all run successful (except windows which takes longer....)

gittiver reviewed (changes requested):
I just notized that documentation needs to be adapted.
